docs/src/testing/gas-etimation.md

```diff
@@ -35,7 +38,30 @@ Multiplication of those values gives us a gas cost for each component:
 
 We should remember that only the most expensive factor will be taken into account, so our overall gas cost is `20.48`.
 
+### From Used Onchain Data
+
+[Starknet documentation](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/fee-mechanism/#_on_chain_data)
+splits onchain data into three parts:
+- `storage updates`
+- `L2 -> L1 messages`
+- `deployed contracts`
+
+Let's calculate gas based on those operations:
+- contract deployment
+- one storage write
+- one L2 -> L1 message (with `[1, 2, 3]` as a payload)
```

Comment:
We could reorder these so it matches the parts above (maybe even explicitly connect them).